### ISSUE TITLE
undefined symbol: X509_VERIFY_PARAM_set1_ip_asc

### ISSUE THREAD

**Issue by agentmario** [certbot/certbot#3405]
Openssl version:OpenSSL 1.0.2h  3 May 2016
OS:Ubuntu 14.04 
Kernel:x64 3.16.0-43
Errlog:
```
Traceback (most recent call last):
  File "/root/.local/share/letsencrypt/bin/letsencrypt", line 7, in <module>
    from certbot.main import main
  File "/root/.local/share/letsencrypt/local/lib/python2.7/site-packages/certbot/main.py", line 14, in <module>
    from acme import jose
  File "/root/.local/share/letsencrypt/local/lib/python2.7/site-packages/acme/jose/__init__.py", line 37, in <module>
    from acme.jose.interfaces import JSONDeSerializable
  File "/root/.local/share/letsencrypt/local/lib/python2.7/site-packages/acme/jose/interfaces.py", line 9, in <module>
    from acme.jose import util
  File "/root/.local/share/letsencrypt/local/lib/python2.7/site-packages/acme/jose/util.py", line 5, in <module>
    import OpenSSL
  File "/root/.local/share/letsencrypt/local/lib/python2.7/site-packages/OpenSSL/__init__.py", line 8, in <module>
    from OpenSSL import rand, crypto, SSL
  File "/root/.local/share/letsencrypt/local/lib/python2.7/site-packages/OpenSSL/rand.py", line 11, in <module>
    from OpenSSL._util import (
  File "/root/.local/share/letsencrypt/local/lib/python2.7/site-packages/OpenSSL/_util.py", line 6, in <module>
    from cryptography.hazmat.bindings.openssl.binding import Binding
  File "/root/.local/share/letsencrypt/local/lib/python2.7/site-packages/cryptography/hazmat/bindings/openssl/binding.py", line 14, in <module>
    from cryptography.hazmat.bindings._openssl import ffi, lib
ImportError: /root/.local/share/letsencrypt/local/lib/python2.7/site-packages/cryptography/hazmat/bindings/_openssl.so: undefined symbol: X509_VERIFY_PARAM_set1_ip_asc
```

**Comment by bmw** [comment]
Can you provide the output of `/root/.local/share/letsencrypt/bin/pip freeze` so we can see the versions of the Python packages being used?

**Comment by agentmario** [comment]
```
 /root/.local/share/letsencrypt/bin/pip freeze
acme==0.8.1
certbot==0.8.1
certbot-apache==0.8.1
cffi==1.4.2
ConfigArgParse==0.10.0
configobj==5.0.6
cryptography==1.2.3
enum34==1.1.2
funcsigs==0.4
idna==2.0
ipaddress==1.0.16
letsencrypt==0.7.0
linecache2==1.0.0
mock==1.0.1
ndg-httpsclient==0.4.0
ordereddict==1.1
parsedatetime==2.1
pbr==1.8.1
psutil==3.3.0
pyasn1==0.1.9
pycparser==2.14
pyOpenSSL==0.15.1
pyRFC3339==1.0
python-augeas==0.5.0
python2-pythondialog==3.3.0
pytz==2015.7
requests==2.9.1
six==1.10.0
traceback2==1.4.0
unittest2==1.1.0
wheel==0.29.0
zope.component==4.2.2
zope.event==4.1.0
zope.interface==4.1.3
```

**Comment by bmw** [comment]
Sorry to ping you again, but any ideas @reaperhulk? `OpenSSL 1.0.2h` with `cryptography 1.2.3` compiles but errors out with the following at runtime:
```
ImportError: /root/.local/share/letsencrypt/local/lib/python2.7/site-packages/cryptography/hazmat/bindings/_openssl.so: undefined symbol: X509_VERIFY_PARAM_set1_ip_asc
```

**Comment by reaperhulk** [comment]
`X509_VERIFY_PARAM_set1_ip_asc` is a symbol added in 1.0.2-beta2. If it's failing at runtime then the balance of probability is that when compiling against 1.0.2 you didn't set it up to properly look for that version of the library. 1.0.2, 1.0.1, and 1.0.0 all have the same soversion by default so the dynamic linker is almost certainly just loading the 1.0.1 version from your default Ubuntu paths. You can probably confirm this by doing an `ldd` on the `_openssl.so`. You'll probably see something like:

```
	libssl.so.1.0.0 => /lib/x86_64-linux-gnu/libssl.so.1.0.0 (0x00007ff647c04000)
	libcrypto.so.1.0.0 => /lib/x86_64-linux-gnu/libcrypto.so.1.0.0 (0x00007ff647828000)
```

When what you really want to see is

```
	libssl.so.1.0.0 => /usr/local/openssl/lib/libssl.so.1.0.0
	libcrypto.so.1.0.0 => /usr/local/openssl/lib/libcrypto.so.1.0.0
```

or whatever the path to your 1.0.2 install is.

It is non-trivial and tricky to build a custom OpenSSL and have libraries link against it successfully in Linux. I recommend against it unless you are familiar with linker args like `rpath` or comfortable with altering soversions.

**Comment by agentmario** [comment]
I've tried to compile openssl 1.1.0 with rpath args and installed. But ldd text was not changed.

**Comment by agentmario** [comment]
Fixed by upgrading OS to 16.04.
